### PR TITLE
Optimize pass-through cases

### PR DESCRIPTION
**Pull request by rightfold**
See #1249.

The following points are still left to do but may be fixed in a later version. They are not required for correct functioning of this optimisation; they just catch more cases:
- Nested patterns (i.e. `case ... of C (D x) -> C (D x)`).
- Multiple expressions being matched against (i.e. `case a, b, c of ...`).
- [ ] Case alternatives involving guards.
- [ ] Find fixed point in CoreFn optimization driver.
- [x] Nested expressions (e.g. `case ... of C x -> f (C x)`).

I doubt the first two cases come up more often than rarely, so I'm leaving them for what they are right now (hence no checkboxes).

The following points may need attention:
- [x] Use `Test.Assert` in the unit test? Not sure whether that's what that module is in tests/support for.
- [x] Suspicious use of `disqualify` and `runProperName`. I am not fully aware of the way names work in psc, so I'll have to figure this out first.

### CODE REVIEW

**Comment by paf31** [comment]
Could you please update `CONTRIBUTORS.md`?

File: src/Language/PureScript/CoreFn/Optimizer.hs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+module Language.PureScript.CoreFn.Optimizer (optimize) where
+
+import Control.Monad.Reader (asks, MonadReader)
+import Language.PureScript.CoreFn.Ann (Ann)
+import Language.PureScript.CoreFn.Module (Module)
+import Language.PureScript.CoreFn.Optimizer.PassThroughCases
+import Language.PureScript.Options (Options, optionsNoOptimizations)
+
+-- |
+-- Apply a series of optimizer passes to CoreFn
```
Comment (paf31) [inline src/Language/PureScript/CoreFn/Optimizer.hs]: Could you please list the optimizations here, even if it's just the one for now?

File: src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs [hunk 2]
```diff
@@ -0,0 +1,81 @@
+-- |
+-- Optimizer step for simplifying "pass-through cases", for example:
```
Comment (paf31) [inline src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs]: Could you please explain here why this optimization is not possible elsewhere (types) and why it is safe?

File: src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  onExpr :: Expr Ann -> Expr Ann
+  onExpr (Case ss ts cs) = Case ss ts (map optimize cs)
+  onExpr e = e
+
+  optimize :: CaseAlternative Ann -> CaseAlternative Ann
+  optimize (CaseAlternative [bndr@(ConstructorBinder bndrAnn _ ctor prms)] (Right body))
+    | isReconstruction ctor prms body =
+        let (_, comments, type_, _) = bndrAnn
+            varAnn = (Nothing, comments, type_, Nothing)
+         in CaseAlternative [NamedBinder bndrAnn v bndr] (Right $ Var varAnn qv)
```
Comment (paf31) [inline src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs]: `qv` might shadow something here. Could you use `MonadSupply`?

File: src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  optimize a = a
+
+  isReconstruction :: Qualified (ProperName 'ConstructorName)
+                   -> [Binder Ann]
+                   -> Expr Ann
+                   -> Bool
+  isReconstruction ctor prms body =
+    case dissectConstruction body of
+      Just (ctor', args) ->
+        Ident (runProperName $ disqualify ctor) == disqualify ctor'
```
Comment (paf31) [inline src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs]: I think you need to compare module names here too.

File: src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  qv :: Qualified Ident
+  qv = Qualified Nothing v
+
+
+-- |
+-- Return the constructor and arguments from what looks like a data constructor
+-- call.
+--
+dissectConstruction :: Expr Ann -> Maybe (Qualified Ident, [Expr Ann])
+dissectConstruction e = (,) <$> findConstructor e <*> findArguments e
```
Comment (paf31) [inline src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs]: I think you can do this in one pass. In fact, I think it exists elsewhere, but I forget the name.

File: tests/support/flattened/utility.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+'use strict';
+
+// module Main
+
+exports.same = function(a) {
```
Comment (paf31) [inline tests/support/flattened/utility.js]: I think this should probably go in `Test.Assert`.

**Comment by paf31** [comment]
Thanks very much for working on this!

File: src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs (hunk 5 above)
Comment (rightfold) [inline src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs]: @paf31 `unApp` works on the JavaScript AST. I see how it's tail-recursive and single-pass though, so I'll use a similar algorithm for `dissectConstruction`. Thanks.

File: src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs (hunk 3 above)
Comment (rightfold) [inline src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs]: Gah, `everywhereOnValues` isn't monadic; this might take some work.

Comment (rightfold) [inline src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs]: Ok, implemented this.

File: tests/support/flattened/utility.js (hunk 6 above)
Comment (rightfold) [inline tests/support/flattened/utility.js]: It's quite a low-level function that I don't think is useful enough to distribute with `Test.Assert`. It's also quite unreliable as more optimisations get added later.

File: src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs (hunk 4 above)
Comment (rightfold) [inline src/Language/PureScript/CoreFn/Optimizer/PassThroughCases.hs]: Good catch.

File: tests/support/flattened/utility.js (hunk 6 above)
Comment (garyb) [inline tests/support/flattened/utility.js]: Agreed. It's testing reference equality, so not a property we'd usually want to care about. There is `refEq` in the prelude for the `Prim` types, but it's not exported.

**Comment by rightfold** [comment]
Ok, I've fixed those points @paf31.

File: src/Language/PureScript/Make.hs [hunk 7]
```diff
@@ -158,16 +158,17 @@ rebuildModule MakeActions{..} externs m@(Module _ _ moduleName _ _) = do
...
   checkExhaustiveModule env' checked
   regrouped <- createBindingGroups moduleName . collapseBindingGroups $ elaborated
   let mod' = Module ss coms moduleName regrouped exps
       corefn = CF.moduleToCoreFn env' mod'
-      [renamed] = renameInModules [corefn]
+  (optCoreFn, nextVar2) <- runSupplyT nextVar1 $ CF.optimize corefn
+  let [renamed] = renameInModules [optCoreFn]
       exts = moduleToExternsFile mod' env'
-  evalSupplyT nextVar . codegen renamed env' . BU8.toString . B.toStrict . encode $ exts
+  evalSupplyT nextVar2 . codegen renamed env' . BU8.toString . B.toStrict . encode $ exts
```
Comment (garyb) [inline src/Language/PureScript/Make.hs]: Very minor thing, but we tend to prime new versions of a value, whereas numbering might indicate semantically similar values that are used in different places.

Comment (rightfold) [inline src/Language/PureScript/Make.hs]: OK!

File: src/Language/PureScript/CoreFn/Traversals.hs [hunk 8]
```diff
@@ -13,34 +14,59 @@ everywhereOnValues :: (Bind a -> Bind a) ->
...
   h' b = h b
 
-  handleCaseAlternative ca =
-    ca { caseAlternativeBinders = map h' (caseAlternativeBinders ca)
-       , caseAlternativeResult = (map (g' *** g') +++ g') (caseAlternativeResult ca)
-       }
+  handleCaseAlternative ca = do
+    caseAlternativeBinders' <- mapM h' (caseAlternativeBinders ca)
+    caseAlternativeResult' <- case caseAlternativeResult ca of
+      Left guardeds -> Left <$> mapM (\(grd, e) -> (,) <$> g' grd <*> g' e) guardeds
```
Comment (garyb) [inline src/Language/PureScript/CoreFn/Traversals.hs]: You could use `pairM` from `Language.PureScript.Traversals` here I think, like we do in the `AST` traversals:

https://github.com/purescript/purescript/blob/5aaf615decfe20a84be051e118260ded4743f2b0/src/Language/PureScript/AST/Traversals.hs#L218-L222

**Comment by rightfold** [comment]
As for this point:

> - Nested expressions (e.g. `case ... of C x -> f (C x)`).

This has an issue with shadowing. While easy to overcome, many optimisations will have to deal with it. Is it an idea to perform a global alpha renaming step across all of non-global bindings in CoreFn before doing any optimisations, such that each binding has a unique name?

**Comment by paf31** [comment]
Actually, if this pass runs after the `Renamer` pass, we might be okay. I didn't think of that.

**Comment by rightfold** [comment]
Does that pass already exist? If it does then I can extend this optimization to recurse into case alternative bodies.

**Comment by paf31** [comment]
Yes, have a look at `Renamer`. I just don't remember the order they run in.

**Comment by rightfold** [comment]
👍🏻

**Comment by rightfold** [comment]
Apparently renamer runs on the AST, not CoreFn, so it obviously runs before this optimization.

**Comment by garyb** [comment]
Could you rebase this please? I'd love to have this in!

**Comment by rightfold** [comment]
@garyb My test case needs an FFI function. Where must I put it? In the past it was in tests/support/flattened but that directory is gone now

**Comment by garyb** [comment]
Just add a .js file with the same name as the .purs test file.

**Comment by paf31** [comment]
What's the status of this? It seems like it was pretty close to complete.

**Comment by paf31** [comment]
Ping

**Comment by rightfold** [comment]
I totally forgot about this, I think it doesn't merge anymore so it needs some changes to avoid merge conflicts, I'll take a look at it tomorrow.

Sent from my iPhone.

**Comment by paf31** [comment]
Closing as potentially abandoned for now, but please do reopen if that's not the case. Thanks!

**Pull request closed without merging**